### DISCUSSION TITLE
Polymorphic Relations not updating in view.

### DISCUSSION THREAD

**Discussion by ArnaudLier** [livewire/livewire#2093] [1. Help] [answered]
Hello!
Here is my Livewire Component:
```php
<?php

namespace App\Http\Livewire;

use Illuminate\Support\Facades\Validator;
use Livewire\Component;

class CreateRatingForm extends Component
{
    /**
     * The component's state.
     *
     * @var array
     */
    public $state = [];

    /**
     * The resource instance.
     *
     * @var \App\Models\Resource
     */
    public $resource;

    /**
     * Mount the component.
     *
     * @param  mixed  $resource
     * @return void
     */
    public function mount($resource)
    {
        $this->resource = $resource;
    }

    /**
     * Add a new rating to a resource.
     *
     * @return void
     */
    public function createRating()
    {
        $this->resetErrorBag();

        Validator::make($this->state, [
            'body' => ['required', 'string', 'min:80', 'max:255'],
            'rating' => ['required', 'integer', 'max:5']
        ])->validateWithBag('createRating');

        $this->resource->rateOnce($this->state['rating'], $this->state['body']);

        $this->state = [];

        $this->resource = $this->resource->fresh();
        $this->resource->load('ratings');

        $this->emit('saved');
    }

    /**
     * Render the component.
     *
     * @return \Illuminate\Contracts\Foundation\Application|\Illuminate\Contracts\View\Factory|\Illuminate\Contracts\View\View
     */
    public function render()
    {
        return view('resources.create-rating-form');
    }
}
```
my model:
```php
    public function ratings()
    {
        return $this->morphMany(Rating::class, 'rateable');
    }
```

and where I list my ratings in my blade view:
```blade
@foreach($resource->ratings->sortByDesc('created_at') as $rating)
    <div class="bg-white shadow overflow-hidden sm:rounded-lg col-span-2" wire:key="{{ $rating->id }}">
       /** ... **/
    </div>
@endforeach
```
My problem is that the ratings are not updating even if the $resource model property is refreshed.
Any ideas?
Thank you.

**Comment by joshhanley** [comment, marked as answer]
@ZeProf2Code double check your resource relations are loaded correctly. 

If on the initial request, you have done `$relations->load('ratings')` this won't be available after the next network request. 
So make sure you call load again inside your `createRating` method.

Hope this helps!

**Reply by ArnaudLier** [reply]
They are already loaded.

**Reply by joshhanley** [reply]
Ok. Have you checked the rating is saving? 

Try and put a `dump($this->resource->toArray())` at the beginning of createRating method and also put a `dd($this->resource->toArray())` at the end of createRating and compare the output see if it has the new rating. 
- If it doesn't then, double check that the rateOnce method is receiving the correct data and working properly.
- If it has, then there is probably something going wrong with how they're being rendered, so it'd be checking first the the view is getting the correct data to loop over.

**Reply by ArnaudLier** [reply]
Yes, the rating is saving successfully. And the view is getting the correct data if I reload the page.

**Reply by joshhanley** [reply]
So I’m just trying make sure I understand the problem correctly, the ratings show up fine on page load. When you add a rating the rating is saving, but then the ratings are not showing up after the new rating has been saved? Or they are still shown but the new rating isn’t showing?

In your view, above the foreach loop can you @dump($resource->ratings) and add a rating and tell me what is being dumped to the screen after the rating has been created?

**Reply by ArnaudLier** [reply]
The ratings are still show but the new rating isn't showing. The dump is not changing at all when I add the rating too.

**Reply by joshhanley** [reply]
Ah ok no worries. Your foreach loop is that in the create-rating-form or is that in a parent component?

If it’s in a parent component can you paste your class and blade file for the parent component? 

Only reason I ask is because your emiting saved, so something needs to be listening for that. Maybe something needs to be added to the listener.

**Reply by ArnaudLier** [reply]
Yes, I'm emitting `saved` for this message:
```
<x-jet-action-message class="mr-3" on="saved">
                    {{ __('Added.') }}
                </x-jet-action-message>
```
and the @foreach is in create-rating-form, which I call like this:
`@livewire('resources.create-rating-form', ['resource' => $resource])`

**Reply by joshhanley** [reply]
Yep makes sense. Can you share your whole blade file for create-rating-form?

**Reply by ArnaudLier** [reply]
Here it is:
```blade
@can('download', $resource)
<div class="bg-white shadow overflow-hidden sm:rounded-lg col-span-2">
    <div class="px-4 py-5 sm:px-6">
        <form wire:submit.prevent="createRating">
            /** **/
        </form>
    </div>
</div>
@endcan
@dump($resource->ratings)
@foreach($resource->ratings->sortByDesc('created_at') as $rating)
    <div class="bg-white shadow overflow-hidden sm:rounded-lg col-span-2" wire:key="rating-{{ $rating->id }}">
        /** **/
    </div>
@endforeach
```

**Reply by joshhanley** [reply]
@ZeProf2Code try wrapping everything inside your blade view in a div. Livewire requires that there is one root element, see https://laravel-livewire.com/docs/2.x/rendering-components#returning-blade
> Make sure your Blade view only has ONE root element.

Let me know how that goes!

**Reply by ArnaudLier** [reply]
Thanks, it works!

**Reply by joshhanley** [reply]
No worries!
